### ISSUE TITLE
vip-manager does not see the etcd keys written by patroni using etcd3 config

### ISSUE THREAD

**Issue by sleewoo** [cybertec-postgresql/vip-manager#71]
Hello.

Does `vip-manager` support `etcd v3` api?

When using `etcd.hosts` with `patroni`, `vip-manager` can see the `etcd` keys and works properly.

But if using `etcd3.hosts` instead, `vip-manager` throws `etcd error: 100: Key not found (/service/pgcluster/leader)`,
though keys are there.

```
# ETCDCTL_API=3 etcdctl --endpoints http://192.168.4.4:2379 get --prefix / | grep leader
/service/pgcluster/leader
/service/pgcluster/optime/leader
```

Related issue:  https://github.com/zalando/patroni/issues/1822

**Comment by sleewoo** [comment]
so, clearly there is no [clientv3](https://github.com/etcd-io/etcd/blob/master/client/v3/client.go#L15) implementation.

any plans to support this?
`v2` sort of deprecated already.

i'm keen to help, though `Go` is like Chinese to me :)

**Comment by markwort** [comment]
Hey, you're right, there is no support for etcd v3 in vip-manager...

Last time I checked the implementation in patroni was still considered "Beta" (it looks like it still is, at least 2.0.1 does not contain any indication that it is any better compared to 2.0.0) and there was quite an abundance of issues with the golang client: https://github.com/etcd-io/etcd/issues/11154 .

That, combined with the lack of any interest so far led me to simply not do anything.

Now somebody has shown interest (you) , so I can give it another try to see if it is now possible to include go.etcd.io/etcd/client/v3 and make use of the new API...But it looks like a workaround with using the library at the point of a specific commit would be needed: https://github.com/etcd-io/etcd/issues/11154#issuecomment-642538186 

In the end, it might be the case that etcd v2 is officially "sort of deprecated" but the slow adoption of other projects that vip-manager depends on (patroni and the golang client of etcd) makes it a bit difficult.
Additionally, AFAIK there is not yet a deadline for removing etcd v2 support.


Maybe on a sidenote: are you already running any substantial patroni clusters with etcd v3? any issues so far? I'd be very interested to hear!

All the best
Julian

**Comment by sleewoo** [comment]
thanks for pointing to the issue,
seems it's far more complicated, hopefully with `etcd v3.5` it will start "normalizing".

not using `v3` anywhere, just planing to build a brand new cluster.
the idea was to save some hair by using `v3` from the start rather than starting with `v2` then upgrade when support dropped.

on the other side, i'm trying solve a problem that did not arrive yet :)
will go with `v2`, should work for couple years for sure.

**Comment by sleewoo** [comment]
thanks for gorgeous contribution to `PostgreSQL` world!
`vip-manager` is such a delicious replacement for `keepalived`!
and `pgwatch2` is such tremendously amazing!
and now i do not need to explain why `PostgreSQL`, there is `postgres-showcase`! :)

**Comment by tpo** [comment]
> so, clearly there is no [clientv3](https://github.com/etcd-io/etcd/blob/master/client/v3/client.go#L15) implementation.
> 
> any plans to support this?
> `v2` sort of deprecated already.
> 
> i'm keen to help, though `Go` is like Chinese to me :)

Just FYI: if you speak any "standard" object oriented language such as C++, Java, C#, Objective-C, etc. then jumping into hacking on vip-manager in Go should not be terribly difficult. You should find your way around and be able to do stuff by reading the code and copy/pasting/editing. Of course you'd need to make sure that you only copy/paste/edit stuff that you understand.

That said I am not a vip-manager maintainer so the vip-maintainers might very well and rightfully disagree with me here...

(but anyway: cool you brought up the topic: I think etcd3 should certainly not drop off vip-manager's radar in the long term)

**Comment by sleewoo** [comment]
yeah, actually started digging in the sources, pretty captivating so far

**Comment by hanneskasp** [comment]
Hello,
are there any news on etcd3 support? I think I just ran into the same issue with etcd 3.5.1

`etcd error: client: response is invalid json. The endpoint is probably not valid etcd cluster endpoint.
`

Thanks,
Hannes

**Comment by markwort** [comment]
Hey @hanneskasp ,
it is still somewhere on the bucket list.
I hope I can tackle this in march.
The etcd 3.5 release including all go dep shenanigans is hopefully useable by then.

In the meantime, I suggest that you enable the V2 API in etcd and connect patroni to it in `etcd` mode, not `etcd3` mode.

Reopeing this to keep track of etcd v3 support.

Best regards
Julian

**Comment by pashagolub** [comment]
Hello. We've implemented etcd v3 protocol support. WIP may be found here: #86 

Testing is highly appreciated! 

Thanks in advance!

**Comment by SrZorro** [comment]
Hi @pashagolub ,

We tested #86 with etcd V3 and works like a charm, previously we where getting the same errors reported in this issue but not anymore with this build

Thanks
